lnutil/dlclib.go

```diff
@@ -0,0 +1,481 @@
+
+	return s
+}
+
+// DlcCalcOracleSignaturePubKey computes the predicted signature s*G
+// it's just R - h(R||m)A
+func DlcCalcOracleSignaturePubKey(msg []byte, oracleA, oracleR [33]byte) ([33]byte, error) {
+	return computePubKey(oracleA, oracleR, msg)
+}
+
+// calculates P = pubR - h(msg, pubR)pubA
+func computePubKey(pubA, pubR [33]byte, msg []byte) ([33]byte, error) {
+	var returnValue [33]byte
+
+	// Hardcode curve
+	curve := btcec.S256()
+
+	A, err := btcec.ParsePubKey(pubA[:], curve)
+	if err != nil {
+		return returnValue, err
+	}
+
+	R, err := btcec.ParsePubKey(pubR[:], curve)
+	if err != nil {
+		return returnValue, err
+	}
+
+	// e = Hash(messageType, oraclePubQ)
+	var hashInput []byte
+	hashInput = append(msg, R.X.Bytes()...)
+	e := chainhash.HashB(hashInput)
+
+	bigE := new(big.Int).SetBytes(e)
+
+	if bigE.Cmp(curve.N) >= 0 {
+		return returnValue, fmt.Errorf("hash of (msg, pubR) too big")
+	}
+
+	// e * B
+	A.X, A.Y = curve.ScalarMult(A.X, A.Y, e)
+
+	A.Y.Neg(A.Y)
+
+	A.Y.Mod(A.Y, curve.P)
+
+	P := new(btcec.PublicKey)
+
+	// add to R
+	P.X, P.Y = curve.Add(A.X, A.Y, R.X, R.Y)
+	copy(returnValue[:], P.SerializeCompressed())
+	return returnValue, nil
+}
+
+// SettlementTx returns the transaction to settle the contract. ours = the one we generate & sign. Theirs (ours = false) = the one they generated, so we can use their sigs
+func SettlementTx(c *DlcContract, d DlcContractDivision, ours bool) (*wire.MsgTx, error) {
+
+	tx := wire.NewMsgTx()
+	// set version 2, for op_csv
+	tx.Version = 2
+
+	tx.AddTxIn(wire.NewTxIn(&c.FundingOutpoint, nil, nil))
+
+	totalFee := int64(1000) // TODO: Calculate
+	feeEach := int64(float64(totalFee) / float64(2))
+	feeOurs := feeEach
+	feeTheirs := feeEach
+	valueOurs := d.ValueOurs
+	// We don't have enough to pay for a fee. We get 0, our contract partner pays the rest of the fee
+	if valueOurs < feeEach {
+		feeOurs = valueOurs
+		valueOurs = 0
+	} else {
+		valueOurs = d.ValueOurs - feeOurs
+	}
+	totalContractValue := c.TheirFundingAmount + c.OurFundingAmount
+	valueTheirs := totalContractValue - d.ValueOurs
+
+	if valueTheirs < feeEach {
+		feeTheirs = valueTheirs
+		valueTheirs = 0
+		feeOurs = totalFee - feeTheirs
+		valueOurs = d.ValueOurs - feeOurs
+	} else {
+		valueTheirs -= feeTheirs
+	}
+
+	var buf bytes.Buffer
+	binary.Write(&buf, binary.BigEndian, uint64(0))
+	binary.Write(&buf, binary.BigEndian, uint64(0))
+	binary.Write(&buf, binary.BigEndian, uint64(0))
+	binary.Write(&buf, binary.BigEndian, d.OracleValue)
+	oracleSigPub, err := DlcCalcOracleSignaturePubKey(buf.Bytes(), c.OracleA, c.OracleR)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ours = the one we generate & sign. Theirs (ours = false) = the one they generated, so we can use their sigs
+	if ours {
+		if valueTheirs > 0 {
+			tx.AddTxOut(DlcOutput(c.TheirPayoutPub, oracleSigPub, c.OurPayoutPub, valueTheirs))
+		}
+
+		if valueOurs > 0 {
+			var ourPayoutPKH [20]byte
+			copy(ourPayoutPKH[:], btcutil.Hash160(c.OurPayoutPub[:]))
```

Comment:
If the PKH only output uses the same pubkey as in the DlcOutput script, that can allow someone to figure out what oracle signature is being used by observing the transaction (actually not this one; they will have to wait until the PKH output is spent, and then they'll be able to figure it out)

So we need to use a separate PayoutPKH.  In lit channels the whole pubkey is given: MyRefundPub and TheirRefundPub.  I don't think you need the actual pubkey, so maybe pubkey hash alone is OK.  Or the whole key, either way works.

Reply:
Actually I think this attack only can occur if both parties broadcast their SettlementTx at the same time, as that would be needed to reveal both my pubkey and the pubkey+oracleKey.

still, that's possible, so it's better to have a separate PKH output key.